envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -124,31 +141,44 @@ class RBACFilterFactory(
         return methodPermissions
     }
 
-    private fun resolveClients(incomingEndpoint: IncomingEndpoint, roles: List<Role>): List<String> {
-        val clients = incomingEndpoint.clients.flatMap { clientOrRole ->
+    private fun resolveClientComposites(incomingEndpoint: IncomingEndpoint, roles: List<Role>): List<ClientComposite> {
+        val clientComposites = incomingEndpoint.clients.flatMap { clientOrRole ->
             roles.find { it.name == clientOrRole }?.clients ?: setOf(clientOrRole)
         }
-        return clients.sorted()
+        return clientComposites.sorted()
```

Comment:
Just a question. Why do we need clients sorted ?

Reply:
We do not, but it is easier to identify when looking at envoy config when things are sorted and not scrambled up.

Reply:
This question leaded me to discover that we introduced a problem when we migrated from Route-based access control to RBAC. We forgot about ensuring, that clients are distinct. Right now, when a client is specified directly in incoming permissions and ALSO is included in some role, it will be duplicated in the resulting list. 
We had a protection against it in the past, but now it is gone: https://github.com/allegro/envoy-control/blob/4ba99a19d7fabeaa577921750e33621e367bf2a4/envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyIngressRoutesFactory.kt#L184

We can achieve sorting and removing duplicates at the sime time with a change like that:
```kotlin
    private fun resolveClientComposites(incomingEndpoint: IncomingEndpoint, roles: List<Role>): Collection<ClientComposite> {
        val clientComposites = incomingEndpoint.clients.flatMap { clientOrRole ->
            roles.find { it.name == clientOrRole }?.clients ?: setOf(clientOrRole)
        }
        return clientComposites.toSortedSet()
    }
```

Reply:
@lukidzi and the sorting is important, because we build a policy name from it joining clients to a string. This policy name becomes a key in the map, which may be updated later with more policies for the same clients.

https://github.com/allegro/envoy-control/blob/49b4953f56d89420599285ac9276aa57871a45f3/envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt#L78-L80